gateway/src/main/java/org/georchestra/gateway/filter/global/ResolveTargetGlobalFilter.java

```diff
@@ -75,32 +79,57 @@ public class ResolveTargetGlobalFilter implements GlobalFilter, Ordered {
         Route route = (Route) exchange.getAttributes().get(GATEWAY_ROUTE_ATTR);
         Objects.requireNonNull(route, "no route matched, filter shouldn't be hit");
 
-        GeorchestraTargetConfig config = resolveTarget(route);
+        GeorchestraTargetConfig targetConfig = resolveTarget(route);
         log.debug("Storing geOrchestra target config for Route {} request context", route.getId());
-        GeorchestraTargetConfig.setTarget(exchange, config);
+        GeorchestraTargetConfig.setTarget(exchange, targetConfig);
         return chain.filter(exchange);
     }
 
     @VisibleForTesting
     @NonNull
     GeorchestraTargetConfig resolveTarget(@NonNull Route route) {
 
-        GeorchestraTargetConfig target = new GeorchestraTargetConfig().headers(config.getDefaultHeaders())
-                .accessRules(config.getGlobalAccessRules());
+        GeorchestraTargetConfig target = new GeorchestraTargetConfig();
 
+        Optional<Service> service = findService(route);
+
+        setAccessRules(target, service);
+        setHeaderMappings(target, service);
+
+        return target;
+    }
+
+    private void setAccessRules(GeorchestraTargetConfig target, Optional<Service> service) {
+        List<RoleBasedAccessRule> globalAccessRules = config.getGlobalAccessRules();
+        var targetAccessRules = service.map(Service::getAccessRules).filter(Objects::nonNull).filter(l -> !l.isEmpty())
+                .orElse(globalAccessRules);
+
+        target.accessRules(targetAccessRules);
+    }
+
+    private void setHeaderMappings(GeorchestraTargetConfig target, Optional<Service> service) {
+        HeaderMappings defaultHeaders = config.getDefaultHeaders();
+        HeaderMappings mergedHeaders = service.flatMap(Service::headers)
+                .map(serviceHeaders -> merge(defaultHeaders, serviceHeaders)).orElse(defaultHeaders);
+
+        target.headers(mergedHeaders);
```

Comment:
so if we define default and a custom headers set for the service, then the result will be to merge default + the custom ones ? I was not sure if we had to replace the default ones by customized ones, but this sounds fine to me, as it sounds flexible enough to manage every cases.

Reply:
Yes I think that was always the idea and it was broken. Like in, defaults is defaults, applies to all. Then you can add   headers for a specific service, or even override the defaults ?